### DISCUSSION TITLE
Possible to export midi with names and other data intact?

### DISCUSSION THREAD

**Discussion by Sy256** [helge17/tuxguitar#85] [General]
When you export midi is it possible to keep the data like names of each track included? Low priority but I notice 2.0 beta did this, so wondering if that's a simple port to bring over. I think it might've remembered tempo changes and markers as well but I don't care as much about that, mostly just the track names, which just seem to be a missing xml tag or something. Right now it seems to remember track instruments well, but not the names.

When I use https://www.fourmilab.ch/webtools/midicsv/ to turn a midi from this tuxguitar into a csv, I notice it lacks a field that 2.0 beta midis have, called 'Title_t' which seems to be what controls the track name

**Comment by helge17** [comment]
I just commited (4b40135096dec1bf56d36b27390e49c97e8eab02) the regarding patches from 2.0beta:

hash | author date | subject
-- | -- | --
759ee747 | 2018-01-11 05:19:18 +0500 | Added track names to output MIDI sequence
3819b890 | 2018-01-11 10:26:57 +0500 | Added convenient toString() on MidiMessage
7f9dc5dd | 2018-01-11 11:27:37 +0500 | Importing track names from MIDI files
4c65d5b9 | 2018-01-12 01:04:17 +0500 | Fixed issue with improper import of MIDIs with named sequence
e9812f99 | 2018-01-12 07:17:42 +0500 | Fixed issue with null sequence name
8e05b8e6 | 2018-01-12 07:42:19 +0500 | Indentation fixes
d964927d | 2018-01-12 17:40:46 +0500 | toString for MidiEvent, improved it for MidiMessage
7bc560ea | 2018-01-12 19:26:40 +0500 | MidiSongReader can now reuse the same channel for multiple tracks
a5a3550a | 2018-01-13 12:00:29 +0500 | Fixed JACK not implementing addTrackName

I tested export & import of midi files (Linux only): The track names and some more data is preserved. midicsv shows the new data, and also Audacity reads the MIDI files, so I think the patches work well.

Please test it as soon as a new release is online.

**Reply by Sy256** [reply]
Great, thanks! Tested 08-09 and the midi export works great for all I checked it for, but I'm having a problem with midi IMPORT where the tempo no matter what it's specified, is stuck at 120 and can't be changed. It can LOOK like its being changed but even if I set it to 300 it'll keep being 120. I don't know what the other stuff means to check it.

On further testing this weird tempo thing only seems to happen on SOME midis and I can't get it to happen repeatedly. Seems like it might be a problem of export though because it doesn't happen to any midis I made not using this tuxguitar in the past. But it doesn't always happen exporting from this tuxguitar either. Here is one where it went wrong, this is supposed to be 148bpm and it is playing at 120bpm no matter what I set it to. It is made from scratch from this tuxguitar.
[JYShiny.mid.zip](https://github.com/helge17/tuxguitar/files/12317689/JYShiny.mid.zip)
